### ISSUE TITLE
Java 8 LocalDate and DateTime compatibility

### ISSUE THREAD

**Issue by krzykrucz** [mangstadt/biweekly#59]
There could be added LocalDateIteratorFactory and DateTimeIteratorFactory as java.time support. In google-rfc-2445 there are already implemented similar iterators for jodatime compatibility.

**Comment by rozhok** [comment]
It would be awesome to add JodaTime/DateTime api support because now I have to mess between my code which uses new apis and old awkward java.time APIs

**Comment by marcelstoer** [comment]
Indeed, having to pass a `java.util.Date` to `vEvent.setDateStart()` makes me shudder. It's tough to figure out how you need to define a vEvent based on `ZonedDateTime` or `OffsetDateTime` so that the calendar event has the correct timestamps and timezone data.

**Comment by LEODPEN** [comment]
> Indeed, having to pass a `java.util.Date` to `vEvent.setDateStart()` makes me shudder. It's tough to figure out how you need to define a vEvent based on `ZonedDateTime` or `OffsetDateTime` so that the calendar event has the correct timestamps and timezone data.

Hi here, I do the transformation like this
```java
        LocalDateTime start = event.getStartl();
        Instant startInstant = start.atZone(zoneId).toInstant();
        vEvent.setDateStart(Date.from(startInstant));
```
Please let me know if there's a better way!

**Comment by pcornelissen** [comment]
Ideally the new date / time classes would be supported directly  ;-)

**Comment by mangstadt** [comment]
Would it break Android compatibility if the new date/time classes were used?

**Comment by AndrewBedscastle** [comment]
Android 8 features are supported without changing the minimum API level for android apps by using desugaring, see:
https://developer.android.com/studio/write/java8-support-table

So quick answer would be: It depends
Long answer would be: No, it's just fine, but you have to add the coreLibraryDesugaring to your build.gradle file

**Comment by mangstadt** [comment]
The fact that timezones are involved complicates this greatly.

iCalendar files usually come with their own timezone definition data. In order to be fully compliant, biweekly would need to convert these VTIMEZONE components into custom `ZoneId` objects so that date values can be represented as `ZonedDateTime` instances. biweekly is currently able to do this using a [custom `java.util.TimeZone` implementation](https://github.com/mangstadt/biweekly/blob/master/src/main/java/biweekly/io/ICalTimeZone.java).

A workaround would be to have the user supply their own TZID-to-ZoneId mapping so that biweekly can use `ZoneId` instances supplied by Java. Of course, this would be a step backward since biweekly is currently able to read the timezone data in VTIMEZONE components. It would also be a major inconvenience to the user, as they would have to know the TZID values that the iCalendar uses ahead of time. (TZID values are simply unique identifiers and can technically be any random string)

```java
File file = new File("event.ics");
ICalenar ical = Biweekly.parse(file)
  .timezone("Eastern Standard Time", ZoneId.of("America/New_York"))
  .go();
```

The [code that generates dates from recurrence rules](https://github.com/mangstadt/biweekly/tree/master/src/main/java/biweekly/util/com/google/ical) would also need to be overhauled, since that uses timezones. biweekly uses the old [google-rfc-2445](https://code.google.com/p/google-rfc-2445/) project for this.

**Comment by runnermann** [comment]
After spending some time implementing this into a multi-threaded server. I removed it for this very reason as well as the need to save an .ics file to the hard drive. Any plans to replace the old Java Date class with Java.ZonedDateTime ?

**Comment by pcornelissen** [comment]
FYI, you don't need the second formatter:


```java
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.params.provider.Arguments.of;

public class FormatterTest {

    public static Stream<Arguments> zdtProvider() {
        return Stream.of(
                of(ZonedDateTime.of(LocalDateTime.of(1977, 9, 20, 16, 12), ZoneId.of("GMT+2"))),
                of(ZonedDateTime.of(LocalDateTime.of(1979, 9, 9, 9, 2), ZoneId.of("GMT+12"))),
                of(ZonedDateTime.of(LocalDateTime.of(2019, 6, 23, 6, 23), ZoneId.systemDefault())),
                of(ZonedDateTime.now(ZoneOffset.systemDefault())),
                of(ZonedDateTime.now(ZoneOffset.UTC))
        );
    }

    @ParameterizedTest
    @MethodSource("zdtProvider")
    void verifyEquality(ZonedDateTime zdtOne) {
        ZoneId zId = ZoneId.systemDefault();
        ZonedDateTime now = ZonedDateTime.now(ZoneOffset.UTC);
        assertThat(from1(zdtOne)).isEqualTo(from2(zdtOne));
    }


    public static String from1(java.time.ZonedDateTime zdt) {
        DateTimeFormatter fmtr1 = DateTimeFormatter.ofPattern("yyyyMMdd'T'hhmmss");

        String str1 = zdt.format(fmtr1);
        String zoneId = zdt.getZone().toString();
        return zoneId + ":" + str1;

    }

    public static String from2(java.time.ZonedDateTime zdt) {
        DateTimeFormatter fmtr1 = DateTimeFormatter.ofPattern("yyyyMMdd");
        DateTimeFormatter fmtr2 = DateTimeFormatter.ofPattern("hhmmss");

        String str1 = zdt.format(fmtr1);
        String str2 = zdt.format(fmtr2);
        String zoneId = zdt.getZone().toString();
        return zoneId + ":" + str1 + "T" + str2;
    }
}

```
